md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1693 @@
+# WebAssembly Specification
+
+0. [Module](#module)
+0. [Validation](#validation)
+0. [Execution](#execution)
+0. [Types](#types)
+0. [Signatures](#signatures)
+0. [Instruction Families](#instruction-families)
+0. [Instructions](#instructions)
+
+
+## Module
+
+TODO: describe the contents of a module
+
+### Positions within a function body
+
+Function bodies contain a sequence of instructions. A *position* within a
+function refers to an element of the sequence, or to an additional
+*return position* which is sequenced after the end of the function body.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control flow stack*, with entries containing:
+      - Either a [type](#types), `any` which satisfies any type check, or `void`
+        which satisfies no type check.
+      - A *depth* value.
+ - A *type stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped
+from it, validation fails.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Type Stack Pop
+
+To perform a pop from the type stack:
+ - If the type stack's size is less than the control flow stack's top's depth
+   value, return `void`.
+ - Otherwise pop a valid from the type stack and return that.
+
+#### Control Flow Stack Type Merge
+
+To merge two control-flow stack type values:
+ - If either type is `any`, the result is the other type.
+ - Otherwise, if the types are the same, the result is that type.
+ - Otherwise, validation fails with a type error.
+
+#### Function Validation Initialization
+
+The current position starts at the first position. The type stack begins empty.
+The control flow stack starts with one entry. This entry's type is the first
+return type of the function, if there are any, or `void` otherwise. Its block
+depth is zero.
+
+#### Function Body Validation
+
+If the current position is the return position, function return validation is
+prompted and validation of the function is thereafter complete.
+
+Otherwise, the instruction at the current position is remembered, the current
+position is incremented to point to the instruction following it, or if there
+are no instructions following it, to the return position. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, the following generic validation is performed:
+
+For each operand type in the instruction's signature, a type is popped from the
+type stack and checked to be the same as the corresponding operand type. Each
+return type of the instruction's signature is then pushed onto the type stack.
+
```

Comment:
I am hoping to focus on the overall approach of this document first, rather than dive into details of validation which are  being discussed elsewhere.